docs/architecture.md

```diff
@@ -0,0 +1,127 @@
+Finally, you must operate and scale the orchestration server and its underlying data store (for example, Cassandra for Temporal), which are both single points of failure for your application.
+
+## Applications and Databases
+
+Each DBOS application server connects to a Postgres database, called the system database.
+This database stores durably stores workflow and step checkpoints, and queue and message state.
+Its schema and tables are documented [here](./explanations/system-tables.md).
+A single Postgres server can host multiple system databases for multiple DBOS applications.
+However, separate applications (meaning separate code bases) should not share a system database.
+
+For example, in this diagram we deploy two DBOS applications, each with three servers.
+Each application has its own isolated system database on the same physical Postgres server, and all of the application's servers connect to that system database.
+
+<img src={require('@site/static/img/architecture/dbos-system-database.png').default} alt="DBOS System Database" width="750" className="custom-img"/>
+
+## How Workflow Recovery Works
+
+To recover workflows from failures, DBOS checkpoints in its system database the input of each workflow and the output of each step.
+When a program executing a workflow fails, crashes, or is interrupted, DBOS uses those checkpoints to recover the workflow from its last completed step.
+Here's how that works:
+
+1. First, DBOS must detect that workflow execution has failed.
+For a single-node application, this is easy: on startup, DBOS looks up and attempts to recover all incomplete (`PENDING`) workflows.
+In a distributed setting, detecting failed workflow execution can be done automatically through services like [DBOS Conductor](#operating-dbos-in-production-with-conductor) or manually using the admin API (more documentation [here](./production/self-hosting/workflow-recovery.md)).
+
+2. Next, DBOS restarts the interrupted workflow from the beginning by calling it with its checkpointed inputs.
+As the workflow re-executes, it checks before executing each step if that step's output is checkpointed in Postgres.
+If there is a checkpoint, the step returns the checkpointed output instead of executing.
+
+3. Eventually, the recovered workflow reaches a step whose output is **not** checkpointed in Postgres.
+It then executes that step normally and proceeds from there, thus **resuming from the last completed step.**
+
+For DBOS to be able to safely recover a workflow, it must satisfy two requirements:
+
+1. The workflow function must be **deterministic**: if called multiple times with the same inputs, it should invoke the same steps with the same inputs in the same order. If you need to perform a non-deterministic operation like accessing the database, calling a third-party API, generating a random number, or getting the local time, you should do it in a step instead of directly in the workflow function.
```

Comment:
```suggestion
1. The workflow function must be **deterministic**: if executed multiple times,  with the same function arguments and step return values, the workflow should invoke the same steps with the same inputs in the same order. If you need to perform any non-deterministic operation like accessing the database, calling a third-party API, generating a random number, or getting the local time, you should do it in a step instead of directly in the workflow function.
```